File: stacked-web/lib/repositories/saved-filter-repository.ts
import type { FilterCriteria, SavedFilter, SavedFilterWithCount } from "@/lib/types/saved-filter";
import { normalizeFilterCriteria } from "@/lib/types/saved-filter";
import type { SupabaseClient } from "@supabase/supabase-js";
import { TaskRepository } from "@/lib/repositories/task-repository";

type SavedFilterRow = {
  id: string;
  name: string;
  color: string | null;
  criteria: unknown;
  sort_order: number;
};

function mapRow(row: SavedFilterRow): SavedFilter {
  return {
    id: row.id,
    name: row.name,
    color: row.color,
    criteria: normalizeFilterCriteria(row.criteria),
    sortOrder: row.sort_order,
  };
}

export class SavedFilterRepository {
  constructor(private client: SupabaseClient) {}

  private async getUserId(): Promise<string> {
    const {
      data: { user },
    } = await this.client.auth.getUser();
    if (!user?.id) throw new Error("Não autenticado");
    return user.id;
  }

  async fetchSavedFilters(): Promise<SavedFilter[]> {
    const { data, error } = await this.client
      .from("saved_filters")
      .select("id, name, color, criteria, sort_order")
      .order("sort_order", { ascending: true })
      .order("created_at", { ascending: true });
    if (error) throw error;
    return (data as SavedFilterRow[]).map(mapRow);
  }

  async fetchSavedFiltersWithCounts(now = new Date()): Promise<SavedFilterWithCount[]> {
    const filters = await this.fetchSavedFilters();
    const taskRepo = new TaskRepository(this.client);
    const counts = await Promise.all(
      filters.map(async (f) => {
        const { pending } = await taskRepo.fetchFilterResults(f.criteria, now);
        return { ...f, pendingCount: pending.length };
      }),
    );
    return counts;
  }

  async createSavedFilter(input: {
    name: string;
    color: string | null;
    criteria: FilterCriteria;
  }): Promise<SavedFilter> {
    const userId = await this.getUserId();
    const { data, error } = await this.client
      .from("saved_filters")
      .insert({
        user_id: userId,
        name: input.name.trim(),
        color: input.color,
        criteria: input.criteria,
      })
      .select("id, name, color, criteria, sort_order")
      .single();
    if (error) throw error;
    return mapRow(data as SavedFilterRow);
  }

  async updateSavedFilter(
    id: string,
    input: { name: string; color: string | null; criteria: FilterCriteria },
  ): Promise<SavedFilter> {
    const { data, error } = await this.client
      .from("saved_filters")
      .update({
        name: input.name.trim(),
        color: input.color,
        criteria: input.criteria,
      })
      .eq("id", id)
      .select("id, name, color, criteria, sort_order")
      .single();
    if (error) throw error;
    return mapRow(data as SavedFilterRow);
  }

  async deleteSavedFilter(id: string): Promise<void> {
    const { error } = await this.client.from("saved_filters").delete().eq("id", id);
    if (error) throw error;
  }

  async fetchSavedFilterById(id: string): Promise<SavedFilter | null> {
    const { data, error } = await this.client
      .from("saved_filters")
      .select("id, name, color, criteria, sort_order")
      .eq("id", id)
      .maybeSingle();
    if (error) throw error;
    return data ? mapRow(data as SavedFilterRow) : null;
  }
}

File: stacked-ios/Stacked/Services/SavedFilterRepository.swift
import Foundation
import Supabase

@MainActor
final class SavedFilterRepository {
  static let shared = SavedFilterRepository()

  private var client: SupabaseClient { SupabaseService.client }
  private init() {}

  func fetchSavedFilters() async throws -> [SavedFilter] {
    let rows: [SavedFilterRowDTO] = try await client
      .from("saved_filters")
      .select("id, name, color, criteria, sort_order")
      .order("sort_order", ascending: true)
      .order("created_at", ascending: true)
      .execute()
      .value
    return rows.map { $0.toModel() }
  }

  func fetchSavedFiltersWithCounts(todayStr: String, weekStr: String) async throws -> [SavedFilterWithCount] {
    let filters = try await fetchSavedFilters()
    var result: [SavedFilterWithCount] = []
    for filter in filters {
      let split = try await TaskRepository.shared.fetchFilterResults(
        filter.criteria,
        todayStr: todayStr,
        weekStr: weekStr
      )
      result.append(SavedFilterWithCount(filter: filter, pendingCount: split.pending.count))
    }
    return result
  }

  func createSavedFilter(name: String, colorHex: String?, criteria: FilterCriteria) async throws -> SavedFilter {
    guard let userId = client.auth.currentUser?.id else {
      throw NSError(domain: "Stacked", code: 401, userInfo: [NSLocalizedDescriptionKey: "Não autenticado"])
    }
    struct Payload: Encodable {
      let user_id: UUID
      let name: String
      let color: String?
      let criteria: FilterCriteria
    }
    let row: SavedFilterRowDTO = try await client.from("saved_filters")
      .insert(Payload(user_id: userId, name: name.trimmingCharacters(in: .whitespacesAndNewlines), color: colorHex, criteria: criteria))
      .select("id, name, color, criteria, sort_order")
      .single()
      .execute()
      .value
    return row.toModel()
  }

  func updateSavedFilter(id: String, name: String, colorHex: String?, criteria: FilterCriteria) async throws -> SavedFilter {
    struct Payload: Encodable {
      let name: String
      let color: String?
      let criteria: FilterCriteria
    }
    let row: SavedFilterRowDTO = try await client.from("saved_filters")
      .update(Payload(name: name.trimmingCharacters(in: .whitespacesAndNewlines), color: colorHex, criteria: criteria))
      .eq("id", value: id)
      .select("id, name, color, criteria, sort_order")
      .single()
      .execute()
      .value
    return row.toModel()
  }

  func deleteSavedFilter(id: String) async throws {
    try await client.from("saved_filters").delete().eq("id", value: id).execute()
  }
}

private struct SavedFilterRowDTO: Decodable {
  let id: String
  let name: String
  let color: String?
  let criteria: FilterCriteria?
  let sort_order: Int

  func toModel() -> SavedFilter {
    SavedFilter(
      id: id,
      name: name,
      colorHex: color,
      criteria: criteria ?? .empty,
      sortOrder: sort_order
    )
  }

  init(from decoder: Decoder) throws {
    let c = try decoder.container(keyedBy: CodingKeys.self)
    if let s = try? c.decode(String.self, forKey: .id) { id = s }
    else { id = try c.decode(UUID.self, forKey: .id).uuidString }
    name = try c.decode(String.self, forKey: .name)
    color = try c.decodeIfPresent(String.self, forKey: .color)
    criteria = try c.decodeIfPresent(FilterCriteria.self, forKey: .criteria)
    sort_order = try c.decodeIfPresent(Int.self, forKey: .sort_order) ?? 0
  }

  private enum CodingKeys: String, CodingKey {
    case id, name, color, criteria, sort_order
  }
}
